pkg/csi/service/vanilla/controller.go

```diff
@@ -1708,14 +1710,136 @@ func (c *controller) createFileVolume(ctx context.Context, req *csi.CreateVolume
+					vcTopologySegmentsMap[c.managers.CnsConfig.Global.VCenterIP],
+					topology.GetSegments())
+			}
+			log.Debugf("Topology accessibility requirements per VC are %+v", vcTopologySegmentsMap)
+		}
+	}
 
+	if topologyRequirement != nil {
+		// Check if topology domains have been provided in the vSphere CSI config secret.
+		// NOTE: We do not support kubernetes.io/hostname as a topology label.
+		if c.managers.CnsConfig.Labels.TopologyCategories == "" && c.managers.CnsConfig.Labels.Zone == "" &&
+			c.managers.CnsConfig.Labels.Region == "" {
+			return nil, csifault.CSIInvalidArgumentFault, logger.LogNewErrorCode(log, codes.InvalidArgument,
+				"topology category names not specified in the vsphere config secret")
+		}
+	}
 	if !volTaskAlreadyRegistered {
 		var createVolumeSpec = common.CreateVolumeSpec{
 			CapacityMB: volSizeMB,
 			Name:       req.Name,
 			ScParams:   scParams,
 			VolumeType: common.FileVolumeType,
 		}
+		if topologyRequirement != nil {
+			var topologySegmentsList []map[string]string
+			for vcHost, topologySegmentsList = range vcTopologySegmentsMap {
+				// Get VC instance.
+				vcenter, err := common.GetVCenterFromVCHost(ctx, c.managers.VcenterManager, vcHost)
+				if err != nil {
+					return nil, csifault.CSIInternalFault, logger.LogNewErrorCodef(log, codes.Internal,
+						"failed to get vCenter instance for host %q. Error: %+v", vcHost, err)
+				}
+				// If Storage policy is given, check if it exists in the VC. If not found, continue to next VC.
+				var storagePolicyID string
+				if scParams.StoragePolicyName != "" {
+					storagePolicyID, err = vcenter.GetStoragePolicyIDByName(ctx, scParams.StoragePolicyName)
+					if err != nil {
+						// TODO: As govmomi doesn't support locale and all error messages are
+						// in English, we are temporarily resorting to a error message check.
+						// In future, we need to change govmomi to throw a NotFound error and catch that instead.
+						errMssgFromPBM := fmt.Sprintf("no pbm profile found with name: %q",
+							scParams.StoragePolicyName)
+						if err.Error() == errMssgFromPBM {
+							errMsg := fmt.Sprintf("Storage policy name %q not found in VC %q",
+								scParams.StoragePolicyName, vcenter.Config.Host)
+							log.Warn(errMsg)
+							combinedErrMssgs = append(combinedErrMssgs, errMsg)
+							continue
+						}
+						return nil, csifault.CSIInternalFault, logger.LogNewErrorCodef(log, codes.Internal,
+							"failed to get policy ID for storage policy name %q. Error: %+v",
+							scParams.StoragePolicyName, err)
+					}
+					log.Infof("Found ID %q for storage policy name %q in vCenter %q", storagePolicyID,
+						scParams.StoragePolicyName, vcenter.Config.Host)
+				}
+
+				// Get shared accessible datastores for matching topology requirement.
+				if commonco.ContainerOrchestratorUtility.IsFSSEnabled(ctx, common.TopologyPreferentialDatastores) {
+					vcenter, err := c.manager.VcenterManager.GetVirtualCenter(ctx, c.manager.VcenterConfig.Host)
```

Comment:
here we are using `c.manager.VcenterConfig.Host` what about multi vCenter config?

Reply:
Addressed in latest changes